### ISSUE TITLE
FinalClass: If you follow the checkstyle hint, the code breaks. 

### ISSUE THREAD

**Issue by hneemann** [checkstyle/checkstyle#4037]
/var/tmp $ javac Foo.java
/var/tmp $ cat Foo.java
```
/**
 * Foo is a simple test.
 */
public final class Foo {

    /**
     * Test.
     */
    private class Test {

        /**
         * Create a new instance.
         */
        private Test() {
        }
    }

    /**
     * Test 2.
     */
    private class Test2 extends Test {
    }

}
```
/var/tmp $ cat digital_checks.xml 
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalClass"/>
    </module>
</module>
```
/var/tmp $ java -Duser.language=en -jar checkstyle-7.6-all.jar -c digital_checks.xml Foo.java
```
Starting audit...
[ERROR] /home/hneemann/temp/checkstyle/Foo.java:9: Class Test should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.
```

---------------

The class `Test` can not be final because `Test2` inherits from `Test`.  There should be no error at all.
If you decare Test as final, the code does not compile anymore.

--------------

**Comment by unrealwork** [comment]
I am on it!

**Comment by romani** [comment]
connection with @unrealwork  is lost, PR is closed as incomplete.
If smb want to continue from scratch or from where unrealwork is finished, please be welcome .

**Comment by Fernal73** [comment]
https://groups.google.com/forum/m/#!msg/checkstyle/Szdb6677DWM/cCAH59EYAAAJ

Another example Java source that exhibits the above scenario.

**Comment by pbludov** [comment]
Will be fixed in #10859
